File: ios/P3Fitness/Services/RoutineImporter.swift
import Foundation

struct ImportedRoutine {
    var name: String
    var exercises: [ImportedExercise]
}

struct ImportedExercise {
    var name: String
    var sets: Int
    var reps: Int
    var muscleGroup: String
    var equipment: String
}

enum ImporterError: LocalizedError {
    case noAPIKey(String)
    case networkError(String)
    case parseError(String)
    case urlFetchError(String)

    var errorDescription: String? {
        switch self {
        case .noAPIKey(let provider): return "No API key set for \(provider). Please add one in Settings."
        case .networkError(let msg): return "Network error: \(msg)"
        case .parseError(let msg): return "Failed to parse LLM response: \(msg)"
        case .urlFetchError(let msg): return "Failed to fetch URL: \(msg)"
        }
    }
}

final class RoutineImporter {

    // MARK: - Public entry points

    static func importFromURL(_ urlString: String, apiKeyManager: APIKeyManager) async throws -> ImportedRoutine {
        guard let url = URL(string: urlString) else {
            throw ImporterError.urlFetchError("Invalid URL: \(urlString)")
        }
        let (data, _) = try await URLSession.shared.data(from: url)
        let html = String(data: data, encoding: .utf8) ?? ""
        let text = stripHTML(html)
        let truncated = String(text.prefix(8000)) // keep tokens manageable
        return try await sendToLLM(content: truncated, apiKeyManager: apiKeyManager)
    }

    static func importFromText(_ text: String, apiKeyManager: APIKeyManager) async throws -> ImportedRoutine {
        return try await sendToLLM(content: text, apiKeyManager: apiKeyManager)
    }

    // MARK: - LLM dispatch

    private static func sendToLLM(content: String, apiKeyManager: APIKeyManager) async throws -> ImportedRoutine {
        let provider = apiKeyManager.selectedProvider
        guard let apiKey = apiKeyManager.load(for: provider), !apiKey.isEmpty else {
            throw ImporterError.noAPIKey(provider)
        }

        let systemPrompt = """
        You are a fitness assistant. Parse the following workout description and output ONLY valid JSON in this exact format:
        {
          "routineName": "Routine Name",
          "exercises": [
            {"name": "Exercise Name", "sets": 3, "reps": 10, "muscleGroup": "Chest", "equipment": "Barbell"}
          ]
        }
        muscleGroup must be one of: Chest, Back, Shoulders, Arms, Legs, Core, Other.
        equipment must be one of: Barbell, Dumbbell, Cable, Machine, Bodyweight.
        Output ONLY the JSON object, no markdown fences, no extra text.
        """

        let jsonText: String
        switch provider {
        case "openai":
            jsonText = try await callOpenAI(apiKey: apiKey, systemPrompt: systemPrompt, userContent: content)
        case "google":
            jsonText = try await callGemini(apiKey: apiKey, systemPrompt: systemPrompt, userContent: content)
        default: // anthropic
            jsonText = try await callAnthropic(apiKey: apiKey, systemPrompt: systemPrompt, userContent: content)
        }

        return try parseRoutineJSON(jsonText)
    }

    // MARK: - Anthropic

    private static func callAnthropic(apiKey: String, systemPrompt: String, userContent: String) async throws -> String {
        let url = URL(string: "https://api.anthropic.com/v1/messages")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")

        let body: [String: Any] = [
            "model": "claude-haiku-4-5-20251001",
            "max_tokens": 1024,
            "system": systemPrompt,
            "messages": [
                ["role": "user", "content": userContent]
            ]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try checkHTTP(response, data: data, provider: "Anthropic")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]],
              let first = content.first,
              let text = first["text"] as? String else {
            throw ImporterError.parseError("Unexpected Anthropic response structure")
        }
        return text
    }

    // MARK: - OpenAI

    private static func callOpenAI(apiKey: String, systemPrompt: String, userContent: String) async throws -> String {
        let url = URL(string: "https://api.openai.com/v1/chat/completions")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        let body: [String: Any] = [
            "model": "gpt-4o-mini",
            "max_tokens": 1024,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user", "content": userContent]
            ]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try checkHTTP(response, data: data, provider: "OpenAI")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = json["choices"] as? [[String: Any]],
              let first = choices.first,
              let message = first["message"] as? [String: Any],
              let text = message["content"] as? String else {
            throw ImporterError.parseError("Unexpected OpenAI response structure")
        }
        return text
    }

    // MARK: - Google Gemini

    private static func callGemini(apiKey: String, systemPrompt: String, userContent: String) async throws -> String {
        let urlString = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=\(apiKey)"
        let url = URL(string: urlString)!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")

        let combinedPrompt = "\(systemPrompt)\n\n\(userContent)"
        let body: [String: Any] = [
            "contents": [
                ["parts": [["text": combinedPrompt]]]
            ],
            "generationConfig": ["maxOutputTokens": 1024]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try checkHTTP(response, data: data, provider: "Google")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let candidates = json["candidates"] as? [[String: Any]],
              let first = candidates.first,
              let content = first["content"] as? [String: Any],
              let parts = content["parts"] as? [[String: Any]],
              let part = parts.first,
              let text = part["text"] as? String else {
            throw ImporterError.parseError("Unexpected Gemini response structure")
        }
        return text
    }

    // MARK: - Parsing

    private static func parseRoutineJSON(_ raw: String) throws -> ImportedRoutine {
        // Strip markdown fences if present
        var cleaned = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleaned.hasPrefix("```") {
            cleaned = cleaned
                .components(separatedBy: "\n")
                .dropFirst()
                .joined(separator: "\n")
            if cleaned.hasSuffix("```") {
                cleaned = String(cleaned.dropLast(3)).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }

        guard let data = cleaned.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ImporterError.parseError("Response is not valid JSON")
        }

        let routineName = json["routineName"] as? String ?? "Imported Routine"
        let rawExercises = json["exercises"] as? [[String: Any]] ?? []

        let exercises: [ImportedExercise] = rawExercises.compactMap { ex in
            guard let name = ex["name"] as? String else { return nil }
            let sets = ex["sets"] as? Int ?? 3
            let reps = ex["reps"] as? Int ?? 10
            let muscleGroup = ex["muscleGroup"] as? String ?? "Other"
            let equipment = ex["equipment"] as? String ?? "Bodyweight"
            return ImportedExercise(name: name, sets: sets, reps: reps, muscleGroup: muscleGroup, equipment: equipment)
        }

        return ImportedRoutine(name: routineName, exercises: exercises)
    }

    // MARK: - Helpers

    private static func checkHTTP(_ response: URLResponse, data: Data, provider: String) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ImporterError.networkError("\(provider) returned HTTP \(http.statusCode): \(body.prefix(200))")
        }
    }

    private static func stripHTML(_ html: String) -> String {
        // Remove script/style blocks first
        var result = html
        let blockTags = ["script", "style", "head", "nav", "footer", "header"]
        for tag in blockTags {
            while let startRange = result.range(of: "<\(tag)", options: .caseInsensitive),
                  let endRange = result.range(of: "</\(tag)>", options: .caseInsensitive) {
                let removeRange = startRange.lowerBound..<endRange.upperBound
                result.removeSubrange(removeRange)
            }
        }
        // Remove all remaining HTML tags
        result = result.replacingOccurrences(of: "<[^>]+>", with: " ", options: .regularExpression)
        // Collapse whitespace
        result = result.components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
        // Decode common HTML entities
        result = result
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&quot;", with: "\"")
        return result
    }
}

File: app/src/main/java/com/fittrack/data/api/RoutineImporter.kt
package com.fittrack.data.api

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class ImportedExercise(
    val name: String,
    val sets: Int,
    val reps: Int,
    val muscleGroup: String = "",
    val equipment: String = ""
)

data class ImportedRoutine(
    val name: String,
    val exercises: List<ImportedExercise>
)

sealed class ImportState {
    data object Idle : ImportState()
    data object Loading : ImportState()
    data class Preview(val routine: ImportedRoutine) : ImportState()
    data class Error(val message: String) : ImportState()
    data object Success : ImportState()
}

class RoutineImporter(private val apiKeyManager: ApiKeyManager) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    fun importFromUrl(url: String, onResult: (ImportState) -> Unit) {
        val provider = apiKeyManager.getSelectedProvider()
        val apiKey = apiKeyManager.getApiKey(provider)
        if (apiKey.isNullOrBlank()) {
            onResult(ImportState.Error("No API key configured for ${provider.displayName}. Add your key in Settings."))
            return
        }

        onResult(ImportState.Loading)

        try {
            val pageRequest = Request.Builder().url(url).build()
            val pageResponse = client.newCall(pageRequest).execute()
            val html = pageResponse.body?.string() ?: ""

            if (html.isBlank()) {
                onResult(ImportState.Error("Could not fetch content from URL"))
                return
            }

            val textContent = html
                .replace(Regex("<script[^>]*>[\\s\\S]*?</script>"), "")
                .replace(Regex("<style[^>]*>[\\s\\S]*?</style>"), "")
                .replace(Regex("<[^>]+>"), " ")
                .replace(Regex("\\s+"), " ")
                .take(8000)

            val routine = callLlmApi(provider, apiKey, buildUrlPrompt(textContent))
            if (routine != null) {
                onResult(ImportState.Preview(routine))
            } else {
                onResult(ImportState.Error("Could not extract a workout routine from this page"))
            }
        } catch (e: Exception) {
            onResult(ImportState.Error(e.message ?: "Import failed"))
        }
    }

    fun importFromText(text: String, onResult: (ImportState) -> Unit) {
        val provider = apiKeyManager.getSelectedProvider()
        val apiKey = apiKeyManager.getApiKey(provider)
        if (apiKey.isNullOrBlank()) {
            onResult(ImportState.Error("No API key configured for ${provider.displayName}. Add your key in Settings."))
            return
        }

        if (text.isBlank()) {
            onResult(ImportState.Error("Please enter a workout description"))
            return
        }

        onResult(ImportState.Loading)

        try {
            val routine = callLlmApi(provider, apiKey, buildTextPrompt(text))
            if (routine != null) {
                onResult(ImportState.Preview(routine))
            } else {
                onResult(ImportState.Error("Could not generate a routine from that description"))
            }
        } catch (e: Exception) {
            onResult(ImportState.Error(e.message ?: "Generation failed"))
        }
    }

    private fun callLlmApi(provider: LlmProvider, apiKey: String, userMessage: String): ImportedRoutine? {
        return when (provider) {
            LlmProvider.ANTHROPIC -> callAnthropicApi(apiKey, userMessage)
            LlmProvider.OPENAI -> callOpenAiApi(apiKey, userMessage)
            LlmProvider.GOOGLE -> callGoogleApi(apiKey, userMessage)
        }
    }

    // ── Anthropic Claude ──

    private fun callAnthropicApi(apiKey: String, userMessage: String): ImportedRoutine? {
        val requestBody = JSONObject().apply {
            put("model", LlmProvider.ANTHROPIC.model)
            put("max_tokens", 2048)
            put("messages", JSONArray().apply {
                put(JSONObject().apply {
                    put("role", "user")
                    put("content", userMessage)
                })
            })
        }

        val request = Request.Builder()
            .url(LlmProvider.ANTHROPIC.apiEndpoint)
            .addHeader("x-api-key", apiKey)
            .addHeader("anthropic-version", "2023-06-01")
            .addHeader("content-type", "application/json")
            .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: return null

        if (!response.isSuccessful) {
            val errorJson = try { JSONObject(responseBody) } catch (_: Exception) { null }
            val errorMsg = errorJson?.optJSONObject("error")?.optString("message") ?: "API error: ${response.code}"
            throw Exception(errorMsg)
        }

        val json = JSONObject(responseBody)
        val content = json.getJSONArray("content")
        val textBlock = (0 until content.length())
            .map { content.getJSONObject(it) }
            .firstOrNull { it.getString("type") == "text" }
            ?.getString("text") ?: return null

        return parseRoutineJson(textBlock)
    }

    // ── OpenAI ──

    private fun callOpenAiApi(apiKey: String, userMessage: String): ImportedRoutine? {
        val requestBody = JSONObject().apply {
            put("model", LlmProvider.OPENAI.model)
            put("max_tokens", 2048)
            put("messages", JSONArray().apply {
                put(JSONObject().apply {
                    put("role", "user")
                    put("content", userMessage)
                })
            })
        }

        val request = Request.Builder()
            .url(LlmProvider.OPENAI.apiEndpoint)
            .addHeader("Authorization", "Bearer $apiKey")
            .addHeader("Content-Type", "application/json")
            .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: return null

        if (!response.isSuccessful) {
            val errorJson = try { JSONObject(responseBody) } catch (_: Exception) { null }
            val errorMsg = errorJson?.optJSONObject("error")?.optString("message") ?: "API error: ${response.code}"
            throw Exception(errorMsg)
        }

        val json = JSONObject(responseBody)
        val choices = json.getJSONArray("choices")
        if (choices.length() == 0) return null
        val text = choices.getJSONObject(0)
            .getJSONObject("message")
            .getString("content")

        return parseRoutineJson(text)
    }

    // ── Google Gemini ──

    private fun callGoogleApi(apiKey: String, userMessage: String): ImportedRoutine? {
        val requestBody = JSONObject().apply {
            put("contents", JSONArray().apply {
                put(JSONObject().apply {
                    put("parts", JSONArray().apply {
                        put(JSONObject().apply {
                            put("text", userMessage)
                        })
                    })
                })
            })
            put("generationConfig", JSONObject().apply {
                put("maxOutputTokens", 2048)
            })
        }

        val url = "${LlmProvider.GOOGLE.apiEndpoint}?key=$apiKey"

        val request = Request.Builder()
            .url(url)
            .addHeader("Content-Type", "application/json")
            .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: return null

        if (!response.isSuccessful) {
            val errorJson = try { JSONObject(responseBody) } catch (_: Exception) { null }
            val errorMsg = errorJson?.optJSONObject("error")?.optString("message") ?: "API error: ${response.code}"
            throw Exception(errorMsg)
        }

        val json = JSONObject(responseBody)
        val candidates = json.getJSONArray("candidates")
        if (candidates.length() == 0) return null
        val text = candidates.getJSONObject(0)
            .getJSONObject("content")
            .getJSONArray("parts")
            .getJSONObject(0)
            .getString("text")

        return parseRoutineJson(text)
    }

    // ── Shared parsing ──

    private fun parseRoutineJson(text: String): ImportedRoutine? {
        val jsonStr = Regex("\\{[\\s\\S]*\\}").find(text)?.value ?: return null

        return try {
            val json = JSONObject(jsonStr)
            val name = json.optString("name", "Imported Routine")
            val exercisesArray = json.getJSONArray("exercises")

            val exercises = (0 until exercisesArray.length()).map { i ->
                val ex = exercisesArray.getJSONObject(i)
                ImportedExercise(
                    name = ex.getString("name"),
                    sets = ex.optInt("sets", 3),
                    reps = ex.optInt("reps", 10),
                    muscleGroup = ex.optString("muscleGroup", ""),
                    equipment = ex.optString("equipment", "")
                )
            }

            ImportedRoutine(name = name, exercises = exercises)
        } catch (_: Exception) {
            null
        }
    }

    private fun buildUrlPrompt(content: String): String = """
Extract the workout routine from this webpage content and return it as JSON.

Return ONLY a JSON object with this exact structure (no other text):
{
  "name": "Routine Name",
  "exercises": [
    {"name": "Exercise Name", "sets": 3, "reps": 10, "muscleGroup": "Chest", "equipment": "Barbell"}
  ]
}

muscleGroup must be one of: Chest, Back, Shoulders, Arms, Legs, Core
equipment must be one of: Barbell, Dumbbell, Cable, Machine, Bodyweight

Webpage content:
$content
""".trimIndent()

    private fun buildTextPrompt(text: String): String = """
Generate a complete workout routine based on this description and return it as JSON.

Return ONLY a JSON object with this exact structure (no other text):
{
  "name": "Routine Name",
  "exercises": [
    {"name": "Exercise Name", "sets": 3, "reps": 10, "muscleGroup": "Chest", "equipment": "Barbell"}
  ]
}

muscleGroup must be one of: Chest, Back, Shoulders, Arms, Legs, Core
equipment must be one of: Barbell, Dumbbell, Cable, Machine, Bodyweight

Choose appropriate sets and reps for each exercise. Include 4-8 exercises.

Description: $text
""".trimIndent()
}
